### ISSUE TITLE
tensorflow only occupies one CPU core out of 4 cores on Pi 3

### ISSUE THREAD

**Issue by qingyangqing** [samjabrahams/tensorflow-on-raspberry-pi#99]
### Describe the Issue
I run an image classification net ('http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz') on Pi. The interesting thing is that only one core is occupied during most of time, except last layer (softmax layer) occupies total 4 cores. 

So, I guess the problem is that threads generated by front layers are too big to be loaded into CPU's shared cache(only 512 KiB) in parallel. I tried to configure intra_op_parallelism_threads parameter to make one single op has more threads, thus I wish they can be loaded into CPU in parallel. 

However, no matter what the number I set for intra_op_parallelism_threads is (up to 128 I ever tried), multiple-cores occupation still only happens in the last softmax layer. 

Do you have any idea about this issue ? 
Thanks !

Information about my Raspberry Pi setup:

* Raspberry Pi model: Pi3
* Operating System used: Raspbian Jessie 
* Version of Python used: 2.7
* SD card memory size: 8G
* Size of USB/other device used as swap (if building from source): 
* TensorFlow git commit hash (if building from source):

**Comment by samjabrahams** [comment]
@qingyangqing thanks for opening this up! It's actually an interesting question, and I don't know the answer off the top of my head. I'll need to take a look at this in more detail when I have the time. I'm thinking we test individual operations (Conv2D, Matmul, etc) and see what their CPU usage profile looks like outside the context of InceptionV3. We want to make sure that parallelism is happening at a basic level before looking at memory bottlenecks as a cause.

If anyone has any other ideas, feel free to try them out and report back here!

**Comment by qingyangqing** [comment]
@samjabrahams I found my mistake... I wish this doesn't waste your time. By my testing, the parallelism exactly occurs at most graph nodes (Matmul, Conv2D, etc). The un-parallelism happens only at the graph warm-up at the first time calling session.run(). After reading your testbenches, I do know I should test my net by more runs.

**Comment by kranthisai** [comment]
@qingyangqing I found the same issue with Object detection api. Raspberry pi is only occupies one CPU core out of 4 cores. I can clearly see from CPU utilization only one core being used here. 

`with detection_graph.as_default():

  with tf.Session(graph=detection_graph) as sess:

    for image_path in TEST_IMAGE_PATHS:

      image = Image.open(image_path)

      # the array based representation of the image will be used later in order to prepare the
      # result image with boxes and labels on it.

      image_np = load_image_into_numpy_array(image)
      # Expand dimensions since the model expects images to have shape: [1, None, None, 3]
      image_np_expanded = np.expand_dims(image_np, axis=0)
      image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')
      # Each box represents a part of the image where a particular object was detected.
      boxes = detection_graph.get_tensor_by_name('detection_boxes:0')
      # Each score represent how level of confidence for each of the objects.
      # Score is shown on the result image, together with the class label.
      scores = detection_graph.get_tensor_by_name('detection_scores:0')
      classes = detection_graph.get_tensor_by_name('detection_classes:0')
      num_detections = detection_graph.get_tensor_by_name('num_detections:0')`
